aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+template <typename scalar_t, int ROW, int COL>
+using array2d = std::array<std::array<scalar_t, COL>, ROW>;
+
+// we consider 6 Taylor expansions of degree
+// 1, 2, 4, 8, 12, 18
+constexpr int total_n_degs = 6;
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+// Allocates a buffers of uninitialized or zero values
+// of shape [n_copies, a.size()]
+Tensor _allocate_buffer(const Tensor& a, int n_copies, bool is_zero = false) {
+  if (is_zero) {
+    return at::zeros(
+      {n_copies, a.size(0), a.size(1), a.size(2)},
+      a.options().memory_format(at::MemoryFormat::Contiguous)
+    );
+  }
+  else {
+    return at::empty(
+      {n_copies, a.size(0), a.size(1), a.size(2)},
+      a.options().memory_format(at::MemoryFormat::Contiguous)
+    );
+  }
+}
+
+// Makes `buffer` to store `num_matrices` number of matrices needed for
+// compute the matrix exponentials of different orders, i.e.
+// first `num_matrices` matrices from the list l := {I, A, A^2, A^3, A^6}
+// in a contiguous block of memory such that
+// buffer[0, ...] = l[0], // I
+// buffer[1, ...] = l[1], // A
+// ...
+// buffer[num_matrices - 1, ...] = l[num_matries - 1]
+void _fill_matrix_powers(Tensor& buffer, const Tensor& a, int num_matrices) {
+  // fill I
+  buffer.select(0, 0).copy_(
+    at::eye(a.size(-1), buffer.options()).expand_as(a)
+  );
+
+  // fill a
+  buffer.select(0, 1).copy_(a);
+
+  // fill a^2
+  if (2 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 2), // out for a^2
+      buffer.select(0, 1),
+      buffer.select(0, 1)
+    );
+  }
+
+  // fill a^3
+  if (3 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 3), // out for a^3
+      buffer.select(0, 1),
+      buffer.select(0, 2)
+    );
+  }
+
+  // fill a^6
+  if (4 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 4),
+      buffer.select(0, 3),
+      buffer.select(0, 3)
+    );
+  }
+}
+
+inline Tensor _pin_and_move_memory_if_cuda_input(
+  const Tensor& mem,
+  const Tensor& in
+) {
+  return (in.device().type() == at::kCUDA)
+    ? mem.pin_memory().to(at::device_of(in).value(), /*non_blocking=*/true)
```

Comment:
What's going on with pinned memory here?

Reply:
@ngimel suggested to pin memory so that the CPU to CUDA move is asynchronous. Can I just use `non_blocking=True` here?

Reply:
I see the use sites of this and I am really skeptical this is helping at all. When you transmit non-pinned host memory to GPU, it is "slower" than pinned memory transfer because CUDA has to internally transfer the memory to pinned memory before actually moving it to CPU. OK, so what does your code do? It... transfers the non-pinned blob to pinned memory, and then moves it to CUDA. This should be entirely a wash. Don't do this.

cc @ngimel

Reply:
Oh, you want `non_blocking`. It's true that non-blocking doesn't do anything if the memory isn't pinned. However, I still don't think it's an appropriate to use non-blocking here. Right after you transfer, you immediately use the data in a kernel. Not very much data-transfer / computation overlapping is going to be possible like that. Plus the data is so small...

Reply:
In the worst case this method transfers the data of size batch_dim.

Reply:
Looks like the performance became even a bit better.